api/v0/ingest/schema/types.go

```diff
@@ -0,0 +1,178 @@
+package schema
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/multiformats/go-multihash"
+)
+
+type (
+	Advertisement struct {
+		PreviousID *ipld.Link
+		Provider   string
+		Addresses  []string
+		Signature  []byte
+		Entries    ipld.Link
+		ContextID  []byte
+		Metadata   []byte
+		IsRm       bool
+	}
+	EntryChunk struct {
+		Entries []multihash.Multihash
+		Next    *ipld.Link
+	}
+
+	// TODO: Propose this to go-ipld-prime as a way to reduce boilerplate code when working with linksystem?
+	//       Loader interface {
+	//       	FromNode(ipld.Node) error
+	//       	LoadFrom(ipld.LinkContext, ipld.LinkSystem, ipld.Link) error
+	//       }
+	//       Storer interface {
+	//       	ToNode() (ipld.Node, error)
+	//       	StoreTo(ipld.LinkContext, ipld.LinkSystem) (ipld.Link, error)
+	//       }
+)
+
+// ToNode converts this advertisement to its representation as an IPLD typed node.
+//
+// See: bindnode.Wrap.
+func (a Advertisement) ToNode() (n ipld.Node, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				err = fmt.Errorf("unknown panic: %v", r)
+			}
+		}
+	}()
+	n = bindnode.Wrap(&a, TypeAdvertisement).Representation()
+	return
+}
+
+// FromNode populates the fields of this advertisement from the given node.
+func (a *Advertisement) FromNode(node ipld.Node) error {
+	if node.Prototype() != AdvertisementPrototype {
+		adBuilder := AdvertisementPrototype.NewBuilder()
+		err := adBuilder.AssignNode(node)
+		if err != nil {
+			return fmt.Errorf("faild to convert node prototype: %w", err)
+		}
+		node = adBuilder.Build()
+	}
+
+	other, ok := bindnode.Unwrap(node).(*Advertisement)
+	if !ok || other == nil {
+		return fmt.Errorf("unwrapped node does not match schema.Advertisement")
+	}
+	a.PreviousID = other.PreviousID
+	a.Provider = other.Provider
+	a.Addresses = other.Addresses
+	a.Signature = other.Signature
+	a.Entries = other.Entries
+	a.ContextID = other.ContextID
+	a.Metadata = other.Metadata
+	a.IsRm = other.IsRm
```

Comment:
is there a particular reason you want to copy the fields rather than returning or using `other` directly?

Reply:
The idea is to have a type-agnostic interface: i.e. `FromNode(ipld.Node) error`.

Reply:
What does that actually gain you, though? Constructors in Go generally don't fit into interfaces, and that's fine. You reach for `bytes.NewBuffer`, not `var buf bytes.Buffer; buf.FromBytes(...)`.

If you want to deduplicate code in the downstream packages, I would suggest having a helper function that works at a higher level, and doing away with the interface entirely. For example:

```
func UnwrapNode(proto schema.TypedPrototype, input ipld.Node) (interface{}, error)
```

This function could do essentially the same as you do now: AssignNode if the prototype doesn't match, otherwise do the Unwrap directly. You do have the `interface{}` return, but you have to deal with bindnode's `interface{}` at some point, given that we can't use generics just yet.

Reply:
> What does that actually gain you, though?

Here it's a total YAGNI no doubt about it :) 

I wanted to avoid casting by keeping the function signature agnostic of the schema types. 
You are right. I'll have a play around to see how much I can remove here.

Reply:
I am torn on this. Here we don't construct, we populate. And that's similar to what a lot of the interfaces in go SDK do, like `ReaderFrom` or `UnmarshalBinary` and other similar encoding interfaces. 

Changing this would also mean I need to get rid of the `LoadFrom` and `StoreTo` receivers on types and have package level helper functions instead. 

Either works. From your perspective, do you still think helper functions for all of it makes more sense?

Reply:
Having looked at the code again I think those interfaces are an overkill here. Ripped them all out, to avoid further bike shedding by yours truly 😊 

Thank you 🍻

Reply:
I'm late to respond here, but methods vs constructors and whether to not use an interface is a bit of a "it depends on the situation".

For example, ReadFrom or UnmarshalBinary do directly modify the receiver in-place. They also need to be an interface, so that custom user types satisfy it the way they want.

The reason I don't think a method fits you is:

1) You are unwrapping into a new value, then copying into the receiver value. So you're doing extra work.
2) You're satisfying an interface that doesn't _really_ need to be there, other than to scratch an itch for abstraction layers :)

Reply:
I should clarify that perhaps an interface would be nice if each type implemented the methods in a different way, but as far as I can see, they would all implement them the same. Which is why I suggest a top-level helper func that does the same for each type, as opposed to each type reimplementing the same logic.

Reply:
Agreed; thank you for your patience with me sir 🍻